File: leetcode/56-merge-intervals.js
// https://leetcode.com/problems/merge-intervals/ (top-interview-questions)

/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
  
  intervals.sort((interval1, interval2) => (interval1[0] - interval2[0]));
  
  const mergedIntervals = [ intervals[0] ];

  for (let i = 1; i < intervals.length; i++) {
    const interval = intervals[i];
    if ( interval[0] <= mergedIntervals[mergedIntervals.length - 1][1] ) {
      mergedIntervals[mergedIntervals.length - 1][1] = Math.max(mergedIntervals[mergedIntervals.length - 1][1], interval[1]);
    } else {
      mergedIntervals.push(interval);
    }
  }

  return mergedIntervals;
};

File: leetcode/56-merge-intervals.py
# https://leetcode.com/problems/merge-intervals/ (top-interview-questions)

from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        
        intervals.sort(key = lambda interval: interval[0])
        output = [intervals[0].copy()]

        for interval in intervals[1:]:
            curr_end = output[-1][1]
            if interval[0] <= curr_end:
                output[-1][1] = max(curr_end, interval[1])
            else:
                output.append(interval.copy())
        
        return output